source/slang/slang-ir-byte-address-legalize.cpp

```diff
@@ -0,0 +1,924 @@
+            // Otherwise, scalar types are assumed
+            // legal for load/store.
+            //
+            return true;
+        }
+
+        // Vector types also depend on the options.
+        //
+        if( as<IRVectorType>(type) )
+        {
+            // If we've been asked to scalarize all
+            // vector load/store, then we need to
+            // tread them as illegal.
+            //
+            if(m_options.scalarizeVectorLoadStore)
+                return false;
+
+        }
+
+        // All other types are treated as always illegal,
+        // so that we will legalize the load/store ops
+        // in all cases.
+        //
+        // Note: recent builds of dxc (perhaps coupled with
+        // recent shader models) support byte-address load/store
+        // of more complex types, but it is simpler for Slang
+        // to just legalize all the composite cases rather
+        // than rely on a downstream compiler.
+        //
+        return false;
+    }
+
+    // The core workhorse routine for the load case is `emitLegalLoad`,
+    // which tries to emit load operations that read a value of the
+    // given `type` from the given `buffer` at the required `baseOffset`
+    // plus the `immediateOffset` if any.
+    //
+    IRInst* emitLegalLoad(IRType* type, IRInst* buffer, IRInst* baseOffset, IRIntegerValue immediateOffset)
+    {
+        // The right way to load a value depends primarily
+        // on the type, and secondarily on the options
+        // that have been specified for this pass.
+        //
+        if( auto structType = as<IRStructType>(type) )
+        {
+            // When loading a value of `struct` type, we will
+            // load each field with its own operation.
+            //
+            // Note: A more "clever" implementation might try
+            // to emit a minimal number of loads of whatever
+            // is the largest supported type matching the
+            // alignment of `structType`, and then break those
+            // loaded values into fields with bit-level ops
+            // once they are in registers.
+            //
+            // Such an approach could conceivably allow more
+            // types to be loadable even on targets that
+            // don't directly support them (e.g., a structure
+            // with an `int` and two `int16_t` could be loadable
+            // even when targetting DXBC).
+            //
+            // The flip side to such an approach would be that
+            // it would complicate the generated code, and also
+            // make the rules about when a type is supported
+            // for byte-address load/store much more complicated.
+
+            // We collect the loaded per-field values into an
+            // array, which we will then use to construct the
+            // full value of the `struct` type.
+            //
+            List<IRInst*> fieldVals;
+            for( auto field : structType->getFields() )
```

Comment:
That I'm assuming we don't need to 'flatten' the structure, because it has already been flattened at this point?